Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -196,109 +171,122 @@ extension SwiftLanguageServer {
   }
 
   // MARK: - Text synchronization
 
-  func openDocument(_ note: Notification<DidOpenTextDocument>) {
+  public func openDocument(_ note: DidOpenTextDocument) {
     guard let snapshot = documentManager.open(note) else {
       // Already logged failure.
       return
     }
 
+    let url = snapshot.document.url
+
+    // Cache the `BuildSystem`'s settings interally.
+    let settings = buildSystem.settings(for: url, snapshot.document.language)
+
     let req = SKRequestDictionary(sourcekitd: sourcekitd)
     req[keys.request] = requests.editor_open
-    req[keys.name] = note.params.textDocument.url.path
+    req[keys.name] = note.textDocument.url.path
     req[keys.sourcetext] = snapshot.text
 
-    // If the BuildSystem has settings, cache them internally.
-    let url = snapshot.document.url
-    if let settings = buildSystem.settings(for: url, snapshot.document.language) {
+    if let settings = settings {
       req[keys.compilerargs] = settings.compilerArguments
-      buildSettingsByFile[url] = settings
-    } else {
-      buildSettingsByFile[url] = nil
     }
 
-    guard let dict = self.sourcekitd.sendSync(req).success else {
-      // Already logged failure.
-      return
-    }
+    // Move to the queue into order to interface with sourcekitd and
+    // update our `BuildSettings.
+    self.queue.async { [weak self] in
+      guard let self = self else { return }
+
+      self.buildSettingsByFile[url] = settings
 
-    publishDiagnostics(response: dict, for: snapshot)
+      guard let dict = self.sourcekitd.sendSync(req).success else {
+        // Already logged failure.
+        return
+      }
+
+      self.publishDiagnostics(response: dict, for: snapshot)
+    }
   }
 
-  func closeDocument(_ note: Notification<DidCloseTextDocument>) {
+  public func closeDocument(_ note: DidCloseTextDocument) {
     documentManager.close(note)
 
-    let url = note.params.textDocument.url
-
-    // Clear settings that should not be cached for closed documents.
-    buildSettingsByFile[url] = nil
-    currentDiagnostics[url] = nil
+    let url = note.textDocument.url
 
     let req = SKRequestDictionary(sourcekitd: sourcekitd)
     req[keys.request] = requests.editor_close
     req[keys.name] = url.path
 
-    _ = self.sourcekitd.sendSync(req)
-  }
+    // Move to the queue into order to interface with sourcekitd and
+    // update our `BuildSettings.
+    self.queue.async { [weak self] in
+      guard let self = self else { return }
 
-  func changeDocument(_ note: Notification<DidChangeTextDocument>) {
+      // Clear settings that should not be cached for closed documents.
+      self.buildSettingsByFile[url] = nil
+      self.currentDiagnostics[url] = nil
 
-    var lastResponse: SKResponseDictionary? = nil
+      _ = self.sourcekitd.sendSync(req)
+    }
+  }
+
+  public func changeDocument(_ note: DidChangeTextDocument) {
+    let req = SKRequestDictionary(sourcekitd: self.sourcekitd)
```

Comment:
The `edit()` call on the document manager will be called once for every edit (it's an array in LSP), but your modified code only works if it is called 0 or 1 times.  I think the original code was okay except that you need to `async` for the `publishDiagnostics` call now.

Reply:
Hmm I think we'll need to `queue.async` for the `self.sourcekitd.sendSync(req).success`, what do you think?

Reply:
I think you can safely make this one `sync` instead of `async`, since we're also using the synchronous sourcekit API here.  That should simplify it.

Reply:
Sure although that will now block the `SourceKitServer`'s queue (where it didn't before)

Reply:
Fair point. In that case it seems like we need to wrap this whole thing in `queue.async`, but move the sourcekitd request back into callback so it gets called the right number of times.

Reply:
Swapped it back